File: Two Sum Problem Python/twosum.py
# Problem Statement:  Given an array of integers and an integer number, return indices of
#                     the two numbers such that they add up to the given target
# Example:
#
# Input: nums = [2,7,11,15], target = 9
# Output: [0,1]
# Explanation: Because nums[0] + nums[1] == 9, we return [0, 1].

# the below code of "TWOSUM" is done in O(nlogn). The left and right pointer perform O(n) steps,
# the time complexity is O(n), but the sorting step makes it O(nlogn).

# the code is ------------------------------------------------------------------------

def two_sum(arr, res):

    #  index of sorted array/list
    a = sorted(range(len(nums)), key=lambda k: arr[k])

    # initialize the left and right
    left = 0
    right = len(arr) - 1

    # two pointers check for a sorted list/array
    while (left < right):

        left = a[left]
        right = a[right]

        if arr[left] + arr[right] == res:
            return [left, right]

        if arr[left] + arr[right] > res:
            right = right - 1
        else:
            left = left + 1

    return []

nums = [2,7,11,15]
target = 17

x = two_sum(nums, target)
print(x)

# answer will be [0, 3] 😊😊😊😊😊😊
# Hope you will understand **************************************************************

File: TwoSum.cpp
#include <bits/stdc++.h>

using namespace std;

vector<int> twoSum(vector<int>& nums, int target) {
        vector<int> ans;
        map<int,int> m;
        int val;
        for(int i=0;i<nums.size();i++)
        {
            val = target - nums[i];
            if(m.count(val))
            {
                ans.push_back(i);
                ans.push_back(m[val]);
            }
            else
                m[nums[i]]=i;
                
        }
        return ans;
    }

int main()
{
    int n;
    cin>>n;
    vector<int> nums(n);
    for(int i=0;i<n;i++)
    cin>>nums[i];
    int target;
    cin>>target;
    vector<int> answer;
    answer = twoSum(nums, target);
    int m=answer.size();
    for(int i=0;i<m;i++)
    cout<<answer[i]<<" ";

    return 0;
}
